network/p2p/scoring/app_score_test.go

```diff
@@ -0,0 +1,295 @@
+	defer p2pfixtures.StopNodes(t, nodes, cancel, 2*time.Second)
+
+	blockTopic := channels.TopicFromChannel(channels.PushBlocks, sporkId)
+	slashingViolationsConsumer := unittest.NetworkSlashingViolationsConsumer(unittest.Logger(), metrics.NewNoopCollector())
+
+	// all nodes subscribe to block topic (common topic among all roles)
+	// group one
+	groupOneSubs := make([]*pubsub.Subscription, len(groupOneNodes))
+	var err error
+	for i, node := range groupOneNodes {
+		groupOneSubs[i], err = node.Subscribe(blockTopic, unittest.NetworkCodec(), unittest.AllowAllPeerFilter(), slashingViolationsConsumer)
+		require.NoError(t, err)
+	}
+	// group two
+	groupTwoSubs := make([]*pubsub.Subscription, len(groupTwoNodes))
+	for i, node := range groupTwoNodes {
+		groupTwoSubs[i], err = node.Subscribe(blockTopic, unittest.NetworkCodec(), unittest.AllowAllPeerFilter(), slashingViolationsConsumer)
+		require.NoError(t, err)
+	}
+	// access node group
+	accessNodeSubs := make([]*pubsub.Subscription, len(accessNodeGroup))
+	for i, node := range accessNodeGroup {
+		accessNodeSubs[i], err = node.Subscribe(blockTopic, unittest.NetworkCodec(), unittest.AllowAllPeerFilter(), slashingViolationsConsumer)
+		require.NoError(t, err)
+	}
+
+	// creates a topology as follows:
+	// groupOneNodes <--> accessNodeGroup <--> groupTwoNodes
+	p2pfixtures.LetNodesDiscoverEachOther(t, ctx, append(groupOneNodes, accessNodeGroup...), append(groupOneIds, accessNodeIds...))
+	p2pfixtures.LetNodesDiscoverEachOther(t, ctx, append(groupTwoNodes, accessNodeGroup...), append(groupTwoIds, accessNodeIds...))
+
+	// checks end-to-end message delivery works
+	// each node sends a distinct message to all and checks that all nodes receive it.
+	for _, node := range nodes {
+		proposalMsg := p2pfixtures.MustEncodeEvent(t, unittest.ProposalFixture())
+		require.NoError(t, node.Publish(ctx, blockTopic, proposalMsg))
+
+		// checks that the message is received by all nodes.
+		ctx1s, cancel1s := context.WithTimeout(ctx, 5*time.Second)
+		p2pfixtures.SubsMustReceiveMessage(t, ctx1s, proposalMsg, groupOneSubs)
+		p2pfixtures.SubsMustReceiveMessage(t, ctx1s, proposalMsg, accessNodeSubs)
+		p2pfixtures.SubsMustReceiveMessage(t, ctx1s, proposalMsg, groupTwoSubs)
+
+		cancel1s()
+	}
+}
+
+// TestNetworkPartitionWithNoHonestPeerScoringInFullTopology is part one of testing pushing access nodes to the edges of the network.
+// This test proves that if access nodes are NOT pushed to the edge of network, a malicious majority of them can
+// partition the network by disconnecting honest nodes from each other even when the network topology is a complete graph (i.e., full topology).
+func TestNetworkPartitionWithNoHonestPeerScoringInFullTopology(t *testing.T) {
+	total := 10
+	partition := 0
+	for i := 0; i < total; i++ {
+		if testNetworkPartitionWithNoHonestPeerScoringInFullTopology(t) {
+			partition++
+		}
+	}
+	require.GreaterOrEqual(t, partition, 1) // at least one partition
```

Comment:
if all we want is at least one, you could end the test as soon as one is detected instead of running through all 10 iterations

Reply:
https://github.com/onflow/flow-go/pull/3379/commits/8b3a828511de2619dfa3a98412adfef79831c43e